src/perf/bin/main.cpp

```diff
@@ -0,0 +1,500 @@
+        if (ServiceHandle == nullptr) {
+            QuicTraceEvent(
+                LibraryErrorStatus,
+                "[ lib] ERROR, %u, %s.",
+                 GetLastError(),
+                "OpenService failed");
+            char DriverFilePath[MAX_PATH];
+            Error =
+                GetFullPathNameA(
+                    "msquictest.sys",
+                    sizeof(DriverFilePath),
+                    DriverFilePath,
+                    nullptr);
+            if (Error == 0) {
+                Error = GetLastError();
+                QuicTraceEvent(
+                    LibraryErrorStatus,
+                    "[ lib] ERROR, %u, %s.",
+                    Error,
+                    "GetFullPathName failed");
+                return false;
+            }
+            ServiceHandle =
+                CreateServiceA(
+                    ScmHandle,
+                    QUIC_PERF_DRIVER_NAME,
+                    QUIC_PERF_DRIVER_NAME,
+                    SC_MANAGER_ALL_ACCESS,
+                    SERVICE_KERNEL_DRIVER,
+                    SERVICE_DEMAND_START,
+                    SERVICE_ERROR_NORMAL,
+                    DriverFilePath,
+                    nullptr,
+                    nullptr,
+                    "msquic\0",
+                    nullptr,
+                    nullptr);
+            if (ServiceHandle == nullptr) {
+                Error = GetLastError();
+                if (Error == ERROR_SERVICE_EXISTS) {
+                    goto QueryService;
+                }
+                QuicTraceEvent(
+                    LibraryErrorStatus,
+                    "[ lib] ERROR, %u, %s.",
+                    Error,
+                    "CreateService failed");
+                return false;
+            }
+        }
+        return true;
+    }
+    ~QuicDriverService() {
+        if (ServiceHandle) {
+            CloseServiceHandle(ServiceHandle);
+        }
+        if (ScmHandle) {
+            CloseServiceHandle(ScmHandle);
+        }
+    }
+    bool Start() {
+        if (!StartServiceA(ServiceHandle, 0, nullptr)) {
+            uint32_t Error = GetLastError();
+            if (Error != ERROR_SERVICE_ALREADY_RUNNING) {
+                QuicTraceEvent(
+                    LibraryErrorStatus,
+                    "[ lib] ERROR, %u, %s.",
+                    Error,
+                    "StartService failed");
+                return false;
+            }
+        }
+        return true;
+    }
+};
+
+#define QUIC_CTL_CODE(request, method, access) \
+    CTL_CODE(FILE_DEVICE_NETWORK, request, method, access)
+
+#define IoGetFunctionCodeFromCtlCode( ControlCode ) (\
+    ( ControlCode >> 2) & 0x00000FFF )
+
+#define IOCTL_QUIC_SEC_CONFIG \
+    QUIC_CTL_CODE(1, METHOD_BUFFERED, FILE_WRITE_DATA)
+
+#define IOCTL_QUIC_RUN_PERF \
+    QUIC_CTL_CODE(2, METHOD_BUFFERED, FILE_WRITE_DATA)
+
+#define IOCTL_QUIC_READ_DATA \
+    QUIC_CTL_CODE(3, METHOD_BUFFERED, FILE_READ_DATA)
```

Comment:
These should be in a shared header with the driver.

Reply:
Done